share/eds/byzantine/share_proof.go

```diff
@@ -76,18 +70,66 @@ func GetProofsForShares(
 	bGetter blockservice.BlockGetter,
 	root cid.Cid,
 	shares [][]byte,
+	axisType rsmt2d.Axis,
 ) ([]*ShareWithProof, error) {
-	proofs := make([]*ShareWithProof, len(shares))
+	sharesWithProofs := make([]*ShareWithProof, len(shares))
 	for index, share := range shares {
 		if share != nil {
 			proof, err := getProofsAt(ctx, bGetter, root, index, len(shares))
 			if err != nil {
 				return nil, err
 			}
-			proofs[index] = proof
+			sharesWithProofs[index] = &ShareWithProof{
+				Share: share,
+				Proof: &proof,
+				Axis:  axisType,
+			}
 		}
 	}
-	return proofs, nil
+	return sharesWithProofs, nil
+}
+
+// getShareWithProof attempts to get a share with proof for the given share. It first tries to get a row proof
+// and if that fails or proof is invalid, it tries to get a column proof.
+func getShareWithProof(
+	ctx context.Context,
+	bGetter blockservice.BlockGetter,
+	dah *share.Root,
+	share share.Share,
+	x, y int,
+) (*ShareWithProof, error) {
+	width := len(dah.RowRoots)
+	// try row proofs
+	root := dah.RowRoots[y]
+	rootCid := ipld.MustCidFromNamespacedSha256(root)
+	proof, err := getProofsAt(ctx, bGetter, rootCid, x, width)
+	if err == nil {
+		shareWithProof := &ShareWithProof{
+			Share: share,
+			Proof: &proof,
+			Axis:  rsmt2d.Row,
+		}
+		if shareWithProof.Validate(rootCid, x, y, width) {
+			return shareWithProof, nil
+		}
```

Comment:
I think we trust our own local proofs, so feels unnecessary.

Reply:
This proof might contain bad encoded data which will make it invalid. So in this case we need orthogonal proof instead

Reply:
From my understanding proof js already valid here as we collected it from local proof data with got and validated before

Reply:
We collected it from traversing the tree from top to the bottom. Validation goes up by recomputing hashes and validating namespacing rules. Collected proof might fail validation for bad encoded data here

Reply:
This validates the correctness of the proof of any data. This does not check for badly encoded erasure-coded data. That price part is done by rsmt2d, so this won't fail for bed erasure coding.

Reply:
Namespacing rules could have been violated.

Reply:
They are validated earlier when we fetch proofs during reconstruction over IPLD

Reply:
Anyway, it does not hurt to have additional validation, so let's keep it the way it is. We can justify it by disk corruption 😉